fem/bounds.hpp

```diff
@@ -0,0 +1,137 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_BOUND
+#define MFEM_BOUND
+
+#include "../config/config.hpp"
+#include "fespace.hpp"
+
+namespace mfem
+{
+
+/** @name Piecewise linear bounds of bases
+    \brief Piecewise linear bounds of bases can be used to compute bounds on the grid function in each element. The bounds for the bases are constructed based on the following parameters:
+
+    (i) @b nb: number of bases/nodes in 1D (i.e. polynomial order+1),
+
+    (ii) @b b_type: bases type, 0 - Lagrange interpolants on Gauss-Legendre nodes, 1 - Lagrange interpolants on Gauss-Lobatto-Legendre nodes, and
+    2 - Positive/Bernstein bases on uniformly distributed nodes,
+
+    (iii) @b ncp: number of control points used to construct the piecewise linear bounds
+
+    (iv) @b cp_type: control point distribution. 0 - GL + end-points,
+                    1 - Chebyshev.
+
+    Note: @b nb and @b b_type are inferred directly from the grid-function.
+
+    If the user does not specify @b ncp and @b cp_type, the minimum value of
+    @b ncp is used that would bound the bases for the @b cp_type. We default
+    to @b cp_type = 0 as it requires fewer number of points to bound the bases. Typically, @b ncp = 2 @b nb is sufficient to get fairly compact bounds, and increasing @b ncp results in tighter bounds.
+
+    Finally, only tensor-product elements are currently supported.
+
+    For more technical details see:
+    Mittal et al., "General Field Evaluation in High-Order Meshes on GPUs" &
+    Dzanic et al., "A method for bounding high-order finite element
+    functions: Applications to mesh validity and bounds-preserving limiters".
+*/
+class PLBound
+{
+private:
+   int nb; // #mesh nodes in 1D
+   int ncp; // #control points in 1D
+   int b_type; // bases type: 0 - GL, 1 - GLL, 2 - Bernstein
+   int cp_type; // control points type: 0 - GL+Ends, 1 - Chebyshev
+   bool proj = true; // Use linear projection to compute bounds.
+   real_t tol = 0.0; // offset bounds to avoid round-off errors
+   Vector nodes, weights, control_points;
+   DenseMatrix lbound, ubound; // nb x ncp matrices with bounds of all bases
+   // Some auxillary storage for computing the bounds with Bernstein
+   DenseMatrix basisMatNodes; // Bernstein bases at equispaced nodes
+   DenseMatrix basisMatInt;   // Bernstein bases at GLL nodes
+   Vector nodes_int, weights_int; // Integration nodes and weights
+   DenseMatrix basisMatLU;    // Used to compute LU factors for Bernstein
+   LUFactors lu;
```

Comment:
I think this should be deleted as member data, and moved to be a local variable where it is used.

The reason is because `LUFactors` is a lightweight object that just contains two pointers. If the `PLBound` object is copied, then `basisMatLU` and `lu_ip` will be copied, and the old pointers will be invalidated, resulting in a segfault.

Because `LUFactors` is so lightweight, there is basically no meaningful overhead to recreating it when you need it (the factorization will still only be done once).